src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,509 @@
+
+void cvsd_device_base::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
 {
-	return (( m_active_clock_hi && !m_last_clock_state &&  clock_state) ||
-			(!m_active_clock_hi &&  m_last_clock_state && !clock_state));
+	// Stub, just return silence
+	auto &buffer = outputs[0];
+
+	m_samples_generated += buffer.samples();
+	if (m_samples_generated >= SAMPLE_RATE)
+		m_samples_generated -= SAMPLE_RATE;
+	for (int i = 0; i < buffer.samples(); i++)
+		buffer.put(i, stream_buffer::sample_t(0.0));
 }
 
 
-inline int hc55516_device::current_clock_state()
+//#########################################
+//                 HC55516
+//#########################################
+DEFINE_DEVICE_TYPE(HC55516, hc55516_device, "hc55516", "HC-55516")
+
+hc55516_device::hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: cvsd_device_base(mconfig, HC55516, tag, owner, clock, RISING, 0x7)
+	, m_agc_push_cb(*this)
+	, m_fzq_pull_cb(*this)
+	, m_sylmask(0xfc0)
+	, m_sylshift(6)
+	, m_syladd(0xfc1)
+	, m_intshift(4)
+	, m_sylfilter(0)
+	, m_intfilter(0)
+	, m_agc(true)
+	, m_buffered_fzq(true)
 {
-	return ((uint64_t)m_update_count * clock() * 2 / SAMPLE_RATE) & 0x01;
 }
 
+// overridable type for hc55532 etc
+hc55516_device::hc55516_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, uint32_t sylmask, int32_t sylshift, int32_t syladd, int32_t intshift)
+	: cvsd_device_base(mconfig, type, tag, owner, clock, RISING, 0x7)
+	, m_agc_push_cb(*this)
+	, m_fzq_pull_cb(*this)
+	, m_sylmask(sylmask)
+	, m_sylshift(sylshift)
+	, m_syladd(syladd)
+	, m_intshift(intshift)
+	, m_sylfilter(0)
+	, m_intfilter(0)
+	, m_agc(true)
+	, m_buffered_fzq(true)
+{
+}
 
-void hc55516_device::process_digit()
+//-------------------------------------------------
+//  device_start - device-specific start
+//-------------------------------------------------
+
+void hc55516_device::device_start()
 {
-	double integrator = m_integrator, temp;
+	cvsd_device_base::device_start();
+	save_item(NAME(m_sylfilter));
+	save_item(NAME(m_intfilter));
+	save_item(NAME(m_agc));
+	save_item(NAME(m_buffered_fzq));
+
+	/* resolve lines */
+	m_agc_push_cb.resolve();
+	m_fzq_pull_cb.resolve();
+}
 
-	/* shift the bit into the shift register */
-	m_shiftreg = (m_shiftreg << 1) | m_digit;
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
 
-	/* move the estimator up or down a step based on the bit */
-	if (m_digit)
-		integrator += m_filter;
-	else
-		integrator -= m_filter;
+void hc55516_device::device_reset()
+{
+	// simulate /FZ having been held for a while
+	m_sylfilter = 0x3f;
+	m_intfilter = 0;
+	m_agc = true;
+	m_buffered_fzq = true; // assuming /FZ was just released and is now high/inactive
+}
```

Comment:
Remember to call the base class `device_reset` at the start of this.

Reply:
fixed.